bpi.syndicate.inc

```diff
@@ -332,149 +454,181 @@ function bpi_preprocess_bpi_search_results(&$variables) {
- * Munges the query string into key/value pairs array, so it is accepted
- * by the l() function.
- * Ex. admin/bpi?key1=value1&key2=value2
+ * @param string $text
+ *   The text on the link.
+ * @param string $field
+ *   The field to sort by.
+ * @param string $default_direction
+ *   The sorting direction to use if not currently sorting of this field.
  *
- * @param string $key
- *   Parameter key.
- * @param string $value
- *   Parameter value.
- * @param string $ignore
- *   Key to be ignored. Used in removing filters.
- *
- * @return array
- *   An array, in structure:
- *   - key: value
+ * @return string
+ *   The link for sorting.
  */
-function _bpi_build_query($key = '', $value = '', $ignore = NULL) {
-  $destination = drupal_get_destination();
-  $current_path = parse_url(reset($destination));
-  $query = isset($current_path['query']) ? $current_path['query'] : '';
-
-  $new_query_pairs = array();
-
-  if (!empty($query)) {
-    // Get the parameters array.
-    $components = explode('&', $query);
-
-    // Iterate and rearrange them.
-    foreach ($components as $component) {
-      $temp = explode('=', $component);
-
-      // Ignore a key from query string.
-      if ($temp[0] == $ignore) {
-        continue;
-      }
-
-      // It's required to decode some values here, since Drupal's url()
-      // encodes it weird, munging the 'space' entity value (%20).
-      $new_query_pairs[$temp[0]] = urldecode($temp[1]);
-    }
+function bpi_get_sort_link($text, $field, $default_direction = 'asc') {
+  $current_sorting = _bpi_get_current_sorting($field);
+  if ($current_sorting) {
+    $sort_direction = $current_sorting == 'asc' ? 'desc' : 'asc';
+  }
+  else {
+    $sort_direction = $default_direction;
   }
+  $query = _bpi_build_query(array(
+    BPI_QUERY_SORT_NAME => array(
+      $field => $sort_direction,
+    ),
+  ), array(BPI_QUERY_SORT_NAME));
+  $image = $current_sorting ? theme('tablesort_indicator', array('style' => $sort_direction)) : '';
+  return l($text . $image, current_path(), array('query' => $query, 'html' => TRUE));
+}
 
-  // Append the new key/value pair to existing ones.
-  if (!empty($key) && !empty($value) && $key != $ignore) {
-    $new_query_pairs[$key] = $value;
+/**
+ * Get current sorting or direction.
+ *
+ * @param string $field
+ *   Field to get sorting direction for.
+ *
+ * @return string|array
+ *   The current sorting (field => direction) or the direction for a specified
+ *   field.
+ */
+function _bpi_get_current_sorting($field = NULL) {
+  $sorting = _bpi_get_query_parameter(BPI_QUERY_SORT_NAME);
+  if ($field) {
+    return isset($sorting[$field]) ? $sorting[$field] : NULL;
   }
+  return $sorting;
+}
 
-  return $new_query_pairs;
+/**
+ * Get current search.
+ *
+ * @return array
+ *   The search.
+ */
+function _bpi_get_current_search_query() {
+  $text = _bpi_get_query_parameter('text', NULL);
+  $amount = _bpi_get_query_parameter('amount');
```

Comment:
No default value for amount, maybe BPI_CONTENT_PER_PAGE. See comments about it above.